src/Forms/GridField/GridFieldLazyLoader.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+namespace SilverStripe\Forms\GridField;
+
+use SilverStripe\Forms\FormField;
+use SilverStripe\Forms\TabSet;
+use SilverStripe\ORM\ArrayList;
+use SilverStripe\ORM\Filterable;
+use SilverStripe\ORM\Limitable;
+use SilverStripe\ORM\SS_List;
+
+/**
+ * GridFieldLazyLoader alters the {@link GridField} behavior to delay rendering of rows until the tab containing the
+ * GridField is selected by the user.
+ *
+ * @see GridField
+ */
+class GridFieldLazyLoader implements GridField_DataManipulator, GridField_HTMLProvider
+{
+
+    /**
+     * @inheritDoc
+     */
+    public function getManipulatedData(GridField $gridField, SS_List $dataList)
+    {
+        // If we are lazy loading an empty the list
+        if ($this->isLazy($gridField)) {
+            if ($dataList instanceof Filterable) {
+                // If our original list can be filtered, filter out all results.
+                $dataList = $dataList->byIDs([-1]);
+            } else {
+                // If not, create an empty list instead.
+                $dataList = ArrayList::create([]);
+            }
+        }
+        return $dataList;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getHTMLFragments($gridField)
+    {
+        $gridField->addExtraClass($this->isLazy($gridField) ?
+            'grid-field-lazy-loadable' :
+            'grid-field-lazy-loaded');
+        return [];
+    }
+
+    /**
+     * Detect if the current request should include results.
+     * @param GridField $gridField
+     * @return bool
+     */
+    private function isLazy(GridField $gridField)
+    {
+        return
+            $gridField->getRequest()->getHeader('X-Pjax') !== 'CurrentField' &&
+            $this->isInTabSet($gridField);
+    }
+
+    /**
+     * Recursively check if $field is inside a TabSet.
+     * @param FormField $field
+     * @return bool
+     */
+    private function isInTabSet(FormField $field)
```

Comment:
Can we please use protected over private?

Reply:
I view this more as an implementation detail. 

This bit of logic is tightly coupled with the front-end JS in admin. If developers try to override `isInTabSet`, it's almost guaranteed to break the front end logic.

My preference would be to keep it private.

Reply:
Suspicious, but ok